### ISSUE TITLE
vulnerability detected by Avast in kernel from OpenHardwareMonitorLib.sys file

### ISSUE THREAD

**Issue by nicolas2k** [openhardwaremonitor/openhardwaremonitor#1557]
Since few time, I've received a vulnerability detected by Avast in kernel from OpenHardwareMonitorLib.sys file on the version  openhardwaremonitor-v0.9.6.zip : 
![image](https://user-images.githubusercontent.com/8255831/218272763-31dbb2b0-0aac-4a88-97a2-5b5630f1b4dc.png)

I must to disable this option but it doesn't correct for security of my computer and I lost speed cpu and temperature.
![image](https://user-images.githubusercontent.com/8255831/218273034-54db1302-d7ee-4e90-a08f-8949a1c35a6d.png)

I use this option in default :
![image](https://user-images.githubusercontent.com/8255831/218272606-608edf47-40c4-48bd-bf1c-dac2e38bbf44.png)

I think, it's better to remove remote function, is it ?

ps: I use under Windows 11 Pro 22H2 [version 10.0.22621.1194]

**Comment by DarthDestroyer** [comment]
it has to be fully allowed and unblocked!

it will use some functions, which could be used in a bad matter too...

**Comment by nicolas2k** [comment]
I've tried to add exclusion but it blocks always ... It seems I've not solution this way of lib.sys

**Comment by nicolas2k** [comment]
a runway maybe

https://eclypsium.com/wp-content/uploads/2019/11/Mother-of-All-Drivers.pdf

https://www.safebreach.com/resources/hp-touchpoint-analytics-dll-search-order-hijacking-potential-abuses-cve-2019-6333/

ps: I use under Windows 11 Pro 22H2 x64 [version 10.0.22621.1194]
...

**Comment by Majikeyric** [comment]
Hi, I just encountered the same problem (with v0.6.0 or v0.9.6) after latest Avast update. I use Windows 7 x64

**Comment by inquartata** [comment]
Same here. Same notification from Avast, unable to unblock, lost CPU data. What would be the best way to resolve this?

**Comment by BenJul79** [comment]
Same for me. 
As I read on links given in a previous conversation, a solution could be to activate "virtualization-based protection of code integrity". I haven't done it yet because I don't know exactly what it is.

**Comment by nicolas2k** [comment]
Hi!

Is it possible to retrieve old version in archives like https://openhardwaremonitor.org/files/openhardwaremonitor-v0.8.0.zip or 0.7.1 or 0.6.0 ? 

Because, I really need because of the gaps in the laptop which suffers the performance of Intel i7 Gen10 ...

Thanks in advance

**Comment by Adgenodux** [comment]
Same problem too after OHM update from 0.9.5 to 0.9.6 1h ago, now Avast refuses it, even when i downgrade to 0.9.5 and exception don't work for it

I've sent the OHM 0.9.6 archive and the link of this thread to Avast for investigation...

**Comment by nicolas2k** [comment]
Hummm, strange, I've tested on different versions, same problem with see picture below :  
<img width="153" alt="image" src="https://user-images.githubusercontent.com/8255831/220871350-612b5ff2-99b1-43c0-9663-3540d95ee3b0.png">

**Comment by nicolas2k** [comment]
@fonder : no body have old versions in archive to test ? thk

**Comment by Adgenodux** [comment]
> Hummm, strange, I've tested on different versions, same problem with see picture below 👍 <img alt="image" width="153" src="https://user-images.githubusercontent.com/8255831/220871350-612b5ff2-99b1-43c0-9663-3540d95ee3b0.png">

yes, it looks like to be an Avast issue...

**Comment by nicolas2k** [comment]
sure ! maybe more secure actually with some verifications drivers in kernel ... (why not !)

**Comment by Adgenodux** [comment]
hardware monitoring needs to go deeper in the machine, so of course, it makes the antivirus nervous due to security risks...

**Comment by Waynarky** [comment]
I'm having the same issue. Please post a resolution if anyone gets one.

**Comment by inquartata** [comment]
I couldn't find a solution so I ended up deactivating the avast feature that disabled ohm. That protection is now off (restart Avast if you cannot find it in the settings. The message tells you what part of the protection it is. Webshield or something...) and got OHM working, but of course disabling protection is not exactly a good idea...nor is it really a solution...

**Comment by Waynarky** [comment]
OK thanks. Yeah, I'm not sure I want it turned off completely, but Asus is needed. I wonder why its triggering an alert in the first place.

**Comment by nicolas2k** [comment]
> 

With less severe safety, I confirm its operation. 
You (@dev) will have to change your method with the Ohm.sys file or find a restraint and necessary library. I'm not expert

**Comment by Adgenodux** [comment]
or wait hopefully Avast fix for false positive...

**Comment by nicolas2k** [comment]
> false positive

I don't think of false positive, there is a vulnerability of the method directly from memory, see my remark (perhaps) above or here : https://github.com/openhardwaremonitor/openhardwaremonitor/issues/1557#issuecomment-1427047891

Also I think it's minor alert but it block to collecte some measures !

**Comment by nicolas2k** [comment]
I use temporary alternative that I made : disable "Block vulnerable Kernel System Drivers (I'm not sure term this sentence) | Bloquer les pilotes de noyau vulnérables" and open this OHM fully and re-enable this blocking ...
I know it doesn't easy ...

**Comment by nex86** [comment]
We are having this problem as well.
We have reasons to not completely disable the protection.  isn't there a way to make seperate exceptions for specific drivers?
I applied rules to all exceptions on the policity on the Avast business console but all of them are being ignored.

Hopefully this gets sorted out ASAP.

**Comment by nicolas2k** [comment]
Hi!

Is it possible to recompile this OHM (minor version) without recreate the file OpenHardwareMonitorLib.sys each opened to avoid detection from Antivirus, to save time ? Because, I didn't found the source OpenHardwareMonitorLib.sys in this GitHub.


Thanks in advance at founder !

**Comment by keiserS** [comment]
Hi,
can someone verify this is legit file and it can be added as an exception file in Avast? (without disabling this security option in Avast). I have checked the folder and didn't found that file at all (Avast must have removed it to quarantine)

Regards,
K.

**Comment by nex86** [comment]
> Hi, can someone verify this is legit file and it can be added as an exception file in Avast? (without disabling this security option in Avast). I have checked the folder and didn't found that file at all (Avast must have removed it to quarantine)
> 
> Regards, K.

Its a .sys file which I already tried to all kind of exceptions including the full path to the file.
Doesn’t work.

**Comment by nicolas2k** [comment]
Me too !

One solution temporary or an alternative is : 
- after boot, just before start OHM, disable this case temporary 
- start OHM application
-  and enable this case, it works until next reboot ...

**Comment by keiserS** [comment]
> Me too !
> 
> One solution temporary or an alternative is :
> 
> * after boot, just before start OHM, disable this case temporary
> * start OHM application
> * and enable this case, it works until next reboot ...

seems then like a bug report for Avast.
Tnx guys for reply.

K.

**Comment by tjsepka** [comment]
The problem is now appearing with AVG Antivirus:

![image](https://user-images.githubusercontent.com/30416148/225779121-791203cc-f517-49c3-b71f-df73974e0120.png)

**Comment by Adgenodux** [comment]
![cpuidhwmonitor](https://user-images.githubusercontent.com/15816585/226111508-3b126636-7bad-4c11-90a1-2b31320f35ca.jpg)

Tested today CPUID HWMonitor V1.50 and works perfectly without warning nor blocking...

So there is a way...

**Comment by nex86** [comment]
> ![cpuidhwmonitor](https://user-images.githubusercontent.com/15816585/226111508-3b126636-7bad-4c11-90a1-2b31320f35ca.jpg)
> 
> Tested today CPUID HWMonitor V1.50 and works perfectly without warning nor blocking...
> 
> So there is a way...

I don't think that HWMonitor uses the same driver/library... 
openhardwaremonitor is a library used for remote monitoring software while hwmonitor can only be used locally.

**Comment by DarthDestroyer** [comment]
It doesnt... And why should you try to compare Open-Source with Closed-Source and a Company behind it?
Makes no sense!
Licensing and Signing costs a lot... If you are willing to pay it, feel free! But you have to understand, its not even close to be compareable...

Set an Exception or get rid of that crappy antivir...
Sorry to say that this harsh, but there is no other way!

**Comment by nex86** [comment]
> Set an Exception or get rid of that crappy antivir... Sorry to say that this harsh, but there is no other way!


Exception can't be set for individual drivers, the protection can only be disabled entirely so its everything or nothing.
And over 100 customers are using the antivir with a running license, its not only a matter of effort to install a new one for everyone but also a matter of cost.

**Comment by DarthDestroyer** [comment]
You didnt got the point!

Example, Windows Defender... You can set an Exception for that sys file and use it!

Second Point you missed, you can sign and license drivers and tools... So it looks like legit applications, that costs money!

It doesnt mean any license for that antivir...

**Comment by PCAssistSoftware** [comment]
Has anyone tried submitting the file to AVG/Avast to analyze to see if they will mark it as safe?

https://www.avg.com/en-gb/false-positive-file-form#pc

https://www.avast.com/en-gb/false-positive-file-form.php#pc

If the driver genuinally doesn't cause damage and the detection is a false detection then letting them analyze it may allow them to determine it is actually okay and solve the problem.

Just a thought....

**Comment by nex86** [comment]
> Example, Windows Defender... You can set an Exception for that sys file and use it!

we're not talking about Windows Defender, we're talking about the self-defense module of Avast and AVG that does NOT provide exceptions for individual kernel drivers.
The exceptions it offers only apply to the anti-virus module and not for the self-defense module.

**Comment by nex86** [comment]
> Has anyone tried submitting the file to AVG/Avast to analyze to see if they will mark it as safe?
> 
> https://www.avg.com/en-gb/false-positive-file-form#pc
> 
> https://www.avast.com/en-gb/false-positive-file-form.php#pc
> 
> If the driver genuinally doesn't cause damage and the detection is a false detection then letting them analyze it may allow them to determine it is actually okay and solve the problem.
> 
> Just a thought....

yes already have, no response yet.

**Comment by PCAssistSoftware** [comment]
> > Has anyone tried submitting the file to AVG/Avast to analyze to see if they will mark it as safe?
> > https://www.avg.com/en-gb/false-positive-file-form#pc
> > https://www.avast.com/en-gb/false-positive-file-form.php#pc
> > If the driver genuinally doesn't cause damage and the detection is a false detection then letting them analyze it may allow them to determine it is actually okay and solve the problem.
> > Just a thought....
> 
> yes already have, no response yet.

Strange not to get a reply - usually AVG reply quite quickly to requests in my experience - then again I am an AVG reseller so perhaps that is why I get responses.  If you don't get a response let me know and I will try submitting it too

**Comment by tjsepka** [comment]
Where did you find the "openhardwaremonitorlib.sys" file that the AV software is flagging?  It's not where it says it is.  I see "openhardwaremonitorlib.dll" in the directory.

**Comment by PCAssistSoftware** [comment]
> Where did you find the "openhardwaremonitorlib.sys" file that the AV software is flagging? It's not where it says it is. I see "openhardwaremonitorlib.dll" in the directory.

It doesn't get created if the AV flags and blocks it - turn off your AV blocking and you will see it

![image](https://user-images.githubusercontent.com/10872799/226148325-e454cb2b-9cb7-46c5-a1ef-d844956d7f32.png)

**Comment by balboa41** [comment]
Hi guys! I am having the same exact issue. I use both ATERA and AVAST AV. So I submitted cases to both ATERA support and AVAST support. They keep blaming each other. AVAST says not to deactivate such protection as we will be vulnerable, while ATERA says that the driver is up-to-date and doesn't represent a vulnerability or risk. So, we are in the middle, without anything to do more than waiting. I really hope that this gets solved ASAP, as my client has about 100 computers and servers complaining daily. It doesn't look good at all.

**Comment by PCAssistSoftware** [comment]
Details from both AVG and Avast on the vulnerability - https://www.cvedetails.com/cve/CVE-2020-14979/

**Comment by tjsepka** [comment]
It appears that the file is created when the program is run and AVG promptly wipes it.  I guess we'll have to wait for a fix.  I was running version 0.8.0 because it is the only version I found that supports graphing.  While it does run, sesnors that I monitor are no longer shown while others are.

![image](https://user-images.githubusercontent.com/30416148/226491776-1d881539-45aa-43e7-a816-a9601f871a4c.png)

**Comment by tjsepka** [comment]
Note that the motherboard sensors are the missing sensors.

**Comment by nicolas2k** [comment]
don't confuse false virus with vulnerability. In this case, it isn't a virus detected but there is a vulnerability under this method of OHMlib.sys as an explanation see (February 12):

https://eclypsium.com/wp-content/uploads/2019/11/Mother-of-All-Drivers.pdf

https://www.safebreach.com/resources/hp-touchpoint-analytics-dll-search-order-hijacking-potential-abuses-cve-2019-6333/

and as said by Darren here : https://www.cvedetails.com/cve/CVE-2020-14979/

So, we cannot add an exception because it doesn't a viral or false virus. 

have a nice day !

**Comment by nicolas2k** [comment]
I'd like this OHM application, I only found this application which is able to use a widget on the desktop, eg :
<img width="134" alt="image" src="https://user-images.githubusercontent.com/8255831/226946795-06e93733-159c-48e1-9c02-8eacba0f6309.png">

**Comment by tjsepka** [comment]
I agree with the assessment above.

**Comment by helonaut** [comment]
We are using Atera on a LOT of devices for remote hardware monitoring purposes. Does this mean we have to switch RMM solution? A CVE score of >7 is worrisome to say the least...

**Comment by nex86** [comment]
> We are using Atera on a LOT of devices for remote hardware monitoring purposes. Does this mean we have to switch RMM solution? A CVE score of >7 is worrisome to say the least...

Atera still works but you can not monitor things like CPU temperature, (maybe CPU usage?), power consumption and more hardware related things.

**Comment by CallumAlbert** [comment]
Another RMM having this issue was/is Syncro. I reported the issue to Syncro and they've now informed me the issue with Avast has been rectified. I no longer appear to be getting the vulnerable driver issue on our machines. I'm unsure as to whether they changed hardware monitoring library or whether they were able to work with Avast to remove the false positive. It might be worth others giving it a go now with the latest definitions and seeing how you go.

**Comment by PatrickSchmidtSE** [comment]
https://www.safebreach.com/resources/hp-touchpoint-analytics-dll-search-order-hijacking-potential-abuses-cve-2019-6333/
HP has this problem since 2019 and before in their collection tool, as it also used the WinRing 0. Anyone with access to a new HP notebook, can check if they simply removed the WinRing0.dll in newer Versions or perhaps has an updated/fixed version! 

Also besides signing the WinRing0, the solution to the CVE should be replacing the method DeviceIOControl one with a "secure" one. This could be done by openhardwaremonitor oder librehardwaremonitor. I wonder why besides signing, no one is creating an imprived WinRing0.dll

**Comment by nex86** [comment]
We got a response from Avast, saying they won't add an exception for the driver and the developers of the monitoring software (Atera / servereye) and/or the driver have to fix it.
![image](https://github.com/openhardwaremonitor/openhardwaremonitor/assets/4693965/f764c6e1-ebfd-448d-a492-36e663523ab4)

**Comment by PatrickSchmidtSE** [comment]
Oh nice... thats really sad . As many RMM tools are using the lib, i wonder if anyone will build a new signed driver version, we all could use again.

**Comment by PatrickSchmidtSE** [comment]
Does anyone has the actual code changes that would be necessary in the LIB and the driver? 

![image](https://github.com/openhardwaremonitor/openhardwaremonitor/assets/9859622/d99df1ae-17fc-4269-877c-10a02b23447b)

Someone at Libre applied the changes above and said it works in no sign mode.


"Applied the changes suggested in the article with the "secure" version of the device object. Compiled that, put it in LHM, rebooted into no-sign kernel driver mode, and it just works." 

So get the WinRing0 code from here : https://github.com/QCute/WinRing0/releases , change it , recompile and integrate in OHM. It would be nice to have a working branch for non-kernel-mode, and then we could find a solution for the signing process.

**Comment by moellerm** [comment]
The last time I checked, the main issue (at least for me and open source projects) preventing any progress in this area is getting a new driver signed.

An attempt to fix this issue was made in this commit:
https://github.com/openhardwaremonitor/openhardwaremonitor/commit/0e435cafc61fe84d429e8a9500d666e34e3de67b

This should restrict access to the driver to privileged accounts once the driver is loaded. Ideally this should be done in the entry code in the driver, as noted in the source code: 
https://github.com/openhardwaremonitor/openhardwaremonitor/blob/09c16898a4a84c36644e7c07baebf1efd299691e/Hardware/KernelDriver.cs#L70
But this would require the resulting driver to get signed again.

I'm not sure if there is a practical way to access (and exploit) the driver from an unprivileged account the way it is loaded in the Open Hardware Monitor. Perhaps there is a way to access it right after starting the service, but before the access gets restricted. The correct way would certainly be to use the IoCreateDeviceSecure in the driver entry code instead of the IoCreateDevice used currently:

https://github.com/openhardwaremonitor/openhardwaremonitor/blob/09c16898a4a84c36644e7c07baebf1efd299691e/External/WinRing0/OpenLibSys.c#L51

In some vulnerability reports it is claimed that the WinRing0 driver allows to read and write arbitrary memory. To my knowledge this is not the case with the version used in the Open Hardware Monitor, as it seems to have been compiled without the option _PHYSICAL_MEMORY_SUPPORT:
https://github.com/openhardwaremonitor/openhardwaremonitor/blob/09c16898a4a84c36644e7c07baebf1efd299691e/External/WinRing0/OpenLibSys.c#L605

Generally, the architecture used in the Open Hardware Monitor for low level access is not ideal. All lower level hardware access should be done in the driver code and expose only a higher level API that fully validates all input. For example, instead of the current API of "read/write IO port XY" it should be "get temperature XY". But this would require frequent updates of the driver code for every change in supported hardware. And as a consequence as well the frequent signing of newly compiled drivers.

**Comment by namazso** [comment]
> In some vulnerability reports it is claimed that the WinRing0 driver allows to read and write arbitrary memory. To my knowledge this is not the case with the version used in the Open Hardware Monitor, as it seems to have been compiled without the option _PHYSICAL_MEMORY_SUPPORT

It allows arbitrary MSR write, which can be abused to overwrite IA32_LSTAR to make the kernel run on user stack. From there on with some ROP, one can achieve kernel code execution, at which point it's a full takeover, equivalent with the access achievable with physical memory mapping. So that option is completely useless from a security perspective.

**Comment by DenisTimofijuk** [comment]
This issue still persists. 

Edition	Windows 10 Pro
Version	22H2
Installed on	‎2023-‎03-‎09
OS build	19045.4170
Experience	Windows Feature Experience Pack 1000.19054.1000.0


![image](https://github.com/openhardwaremonitor/openhardwaremonitor/assets/34959206/d8ec7e27-a617-4928-a96a-0d20cfc67f79)

**Comment by PatrickSchmidtSE** [comment]
Yes, unless someone renews the driver certificate ( one time ) via buying a required certificate and make the microsoft hardware driver check ( to get the microsoft certificate), this will persist.

**Comment by DrOuille9** [comment]
May be same problem as OpenRGB ?
https://gitlab.com/CalcProgrammer1/OpenRGB/-/issues/2227

**Comment by D1mar0n** [comment]
The driver version here is [1.2.0.5](https://github.com/alex-dow/winRing0-rs). Unfortunately, after replacing it with this version, I still get the same error.

**Comment by Zorono** [comment]
same here... 
![image](https://github.com/user-attachments/assets/d0e9b479-5afd-4249-b4ec-4d5fe3e1dc91)

**Comment by pw-in-irvine** [comment]
Same issue in **Norton 360**.
Submitted to Norton for analysis. No response as of 12/16/2024.

![Screenshot (62)](https://github.com/user-attachments/assets/91714c9d-454d-4b1c-be83-3c64e8e6bd7d)

**Comment by Neustradamus** [comment]
Dear @openhardwaremonitor team,

I wish you a Happy New Year 2025!

Have you progressed on it?

**Comment by IanEarnest** [comment]
In the past 24 hours there has been an issue with running the program - as listed here https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/issues/984

Windows is showing a virus warning of "HackTool:Win32/Winring0"

**Comment by justmaiklas** [comment]
![Image](https://github.com/user-attachments/assets/41234347-5d3a-4723-b8a1-2ed43083d666)

**Comment by RickyCT** [comment]
Same issue here, Windows Defender has now also recognised this as a virus and stopping it from running 😭

**Comment by gitbirb** [comment]
> Same issue here, Windows Defender has now also recognised this as a virus and stopping it from running 😭

Just happened to me too, it detected it as

<img width="391" height="177" alt="Image" src="https://github.com/user-attachments/assets/982dc24f-843c-47bf-880c-d7d82a78ed23" />

**Comment by douglasjunior** [comment]
Same here on Windows Defender (Windows 11)

**Comment by Cybernetic-Fox** [comment]
Yeah, just got that this morning

**Comment by NilsDargent** [comment]
Same here with Windows Defender (Windows 11), and I was extremely worried, so I hope it's nothing. 
I don't use this, I didn't even know it existence before so I don't know what to do.

<img width="729" height="587" alt="Image" src="https://github.com/user-attachments/assets/e6dd9ad4-51fd-485d-af31-d72948b904c5" />

**Comment by josefelipecenaqueleite** [comment]
May be has any one at here on this chat with free recovery key from avast .thats a question. I need a lot  ? send private msg to me
